bundles/org.openhab.core.storage.json/src/main/java/org/openhab/core/storage/json/internal/JsonStorage.java

```diff
@@ -210,20 +215,34 @@ public boolean containsKey(final String key) {
         }
 
         try {
+            String entityClassName = entry.getEntityClassName();
+            JsonElement entityValue = (JsonElement) entry.getValue();
+
+            TypeMigrator migrator = typeMigrators.get(entityClassName);
+            if (migrator != null) {
+                entityClassName = migrator.getNewType();
+                entityValue = migrator.migrate(entityValue);
```

Comment:
Did you also consider migrating the whole storage when it is loaded? With this implementation it looks like it will migrate deserialized values that are never persisted forever?

It might also be easier for users to manipulate the .json file with other tools when it is fully migrated.

Doing the migration immediately may also make it easier to spot migration issues due to `TypeMigrationException`s.

Reply:
Yes, I thought about that. But that would require iterating over the whole storage each time it is opened, which is probably not good for performance. 

Maybe it would be better to `put` the migrated value to the storage (if a migration was necessary).

Reply:
I implemented storing the migrated value after the type migration. This keeps the benefits from gradual migration and also migrate only once. For most storages this will be a complete migration (because all storages used in managed provider will be read in total at startup).